pkg/tenant/invites.go

```diff
@@ -0,0 +1,149 @@
+	}
+	c.JSON(http.StatusOK, out)
+}
+
+// InviteAccept is an authenticated endpoint to accept an invitation to join an
+// organization. The invitation token must be provided in the request body, and the
+// email in the user claims must match the email address in the token. If the
+// invitation is invalid this endpoint returns a 404. If successful, the user is logged
+// into the organization and credentials are set as cookies. Frontends should use this
+// endpoint when a user is already logged in and is accepting an invitation. If the
+// user is not logged in, the Login endpoint should be used instead.
+//
+// Route: /invites/accept
+func (s *Server) InviteAccept(c *gin.Context) {
+	var (
+		ctx context.Context
+		req *api.MemberInviteToken
+		err error
+	)
+
+	// User credentials are required for the Quarterdeck request
+	if ctx, err = middleware.ContextFromRequest(c); err != nil {
+		sentry.Error(c).Err(err).Msg("could not get user credentials from authenticated request")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+	}
+
+	// Parse the token in the request body
+	if err = c.BindJSON(&req); err != nil {
+		sentry.Warn(c).Err(err).Msg("could not parse accept invite token request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(responses.ErrTryLoginAgain))
+		return
+	}
+
+	if req.Token == "" {
+		sentry.Warn(c).Msg("missing token in accept invite request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(responses.ErrTryLoginAgain))
+		return
+	}
+
+	// Create the Quarterdeck request with the token
+	acceptRequest := &qd.UserInviteToken{
+		Token: req.Token,
+	}
+
+	// Call Quarterdeck to accept the invite
+	var rep *qd.LoginReply
+	if rep, err = s.quarterdeck.InviteAccept(ctx, acceptRequest); err != nil {
+		sentry.Debug(c).Err(err).Msg("tracing quarterdeck error in tenant")
+		api.ReplyQuarterdeckError(c, err)
+		return
+	}
+
+	// Parse the new claims from the access token
+	var newClaims *tokens.Claims
+	if newClaims, err = tokens.ParseUnverifiedTokenClaims(rep.AccessToken); err != nil {
+		sentry.Error(c).Err(err).Msg("could not parse claims from newly issued access token")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return
+	}
+
+	var orgID ulid.ULID
+	if orgID = newClaims.ParseOrgID(); ulids.IsZero(orgID) {
+		sentry.Error(c).Msg("could not parse orgID from newly issued claims")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return
+	}
+
+	var userID ulid.ULID
+	if userID = newClaims.ParseUserID(); ulids.IsZero(userID) {
+		sentry.Error(c).Msg("could not parse userID from newly issued claims")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return
+	}
+
+	// Retrieve member from the database to update the invitation status
+	var member *db.Member
+	if member, err = db.RetrieveMember(ctx, orgID, userID); err != nil {
```

Comment:
So if the member record is not in the database and we error here, it might create a situation where the user is in the Quarterdeck org but not in trtl. In that case, should we create a new member here or deal with it manually?

Reply:
For now, let's deal with it manually; the way that would happen is if something in the invitation process went wrong, right? I'd rather know about upstream errors for right now.